test/app/src/testApp.js

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (c) 2018, Okta, Inc. and/or its affiliates. All rights reserved.
+ * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+/* global process, window, document */
+/* eslint-disable no-console */
+import OktaAuth from '@okta/okta-auth-js';
+
+/* eslint-disable prefer-destructuring */
+const DOMAIN = process.env.DOMAIN;
+const CLIENT_ID = process.env.CLIENT_ID;
+/* eslint-enable prefer-destructuring */
+
+const ISSUER = `https://${DOMAIN}/oauth2/default`;
+const HOST = window.location.host;
+const REDIRECT_URI = `http://${HOST}/implicit/callback`;
+
+const oktaAuth = new OktaAuth({
+  // maxClockSkew: Number.MAX_SAFE_INTEGER, // FORCES TOKEN RENEW
+  issuer: ISSUER,
+  clientId: CLIENT_ID,
+  redirectUri: REDIRECT_URI,
+});
+
+function saveCodeVerifier(val) {
+  sessionStorage.setItem('okta-code-verifier', val);
+}
+
+function loadCodeVerifier() {
+  return sessionStorage.getItem('okta-code-verifier');
+}
+
+window.login = async function login(event) {
+  event.preventDefault(); // Necessary to prevent default navigation for redirect below
+
+  const scopes = ['openid', 'email', 'profile'];
+  const codeVerifier = oktaAuth.pkce.generateVerifier();
+  const codeChallenge = await oktaAuth.pkce.computeChallenge(codeVerifier);
+  const responseType = 'code';
+  const responseMode = 'fragment';
+
+  // We must save the code verifier somewhere that will survive the redirect
+  saveCodeVerifier(codeVerifier);
+
+  oktaAuth.token.getWithRedirect({
+    scopes,
+    codeChallenge,
+    responseType,
+    responseMode
+  });
+};
+
+window.logout = async function logout() {
+  oktaAuth.tokenManager.clear();
+  await oktaAuth.signOut();
+  window.location.reload();
+};
+
+async function handleAuthentication() {
+  // The authorization code is in the URL fragment
+  let { authorizationCode } = await oktaAuth.token.parseFromUrl();
+  // The code verifier we have stored locally
```

Comment:
The `state` also needs to be [validated here](https://tools.ietf.org/html/rfc6749#section-4.1.2), assuming it exists.

Reply:
I thought state was validated internal to the library? Is it up to a user of the okta-auth-js library to (also?) validate state?

Reply:
The `state` param appears to only be validated for non-redirect flows using a [post-message callback handler](https://github.com/okta/okta-auth-js/blob/master/lib/token.js#L87). For redirect flows, the `state` param must be read from the `okta-oauth-redirect-params` cookie (if it exists) and validated by the client.

Reply:
state is validated in parseFromUrl(), same as implicit flow.  PKCE adds an *additional* request during this step to validate the code.